core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java

```diff
@@ -0,0 +1,690 @@
+        if (entry.getValue().isRelative) {
+          location = tablePath + CarbonCommonConstants.FILE_SEPARATOR + location;
+        }
+        if (entry.getValue().status.equals(SegmentStatus.SUCCESS.getMessage())) {
+          for (String indexFile : entry.getValue().getFiles()) {
+            indexFiles.put(location + CarbonCommonConstants.FILE_SEPARATOR + indexFile,
+                entry.getValue().mergeFileName);
+          }
+        }
+      }
+    }
+    return indexFiles;
+  }
+
+  /**
+   * Drops the partition related files from the segment file of the segment and writes
+   * to a new file. First iterator over segment file and check the path it needs to be dropped.
+   * And update the status with delete if it found.
+   *
+   * @param uniqueId
+   * @throws IOException
+   */
+  public void dropPartitions(String tablePath, Segment segment, List<PartitionSpec> partitionSpecs,
+      String uniqueId, List<String> toBeDeletedSegments, List<String> toBeUpdatedSegments)
+      throws IOException {
+    readSegment(tablePath, segment.getSegmentFileName());
+    boolean updateSegment = false;
+    for (Map.Entry<String, FolderDetails> entry : segmentFile.getLocationMap().entrySet()) {
+      String location = entry.getKey();
+      if (entry.getValue().isRelative) {
+        location = tablePath + CarbonCommonConstants.FILE_SEPARATOR + location;
+      }
+      Path path = new Path(location);
+      // Update the status to delete if path equals
+      for (PartitionSpec spec : partitionSpecs) {
+        if (path.equals(spec.getLocation())) {
+          entry.getValue().setStatus(SegmentStatus.MARKED_FOR_DELETE.getMessage());
+          updateSegment = true;
+          break;
+        }
+      }
+    }
+    String writePath = CarbonTablePath.getSegmentFilesLocation(tablePath);
+    writePath =
+        writePath + CarbonCommonConstants.FILE_SEPARATOR + segment.getSegmentId() + "_" + uniqueId
+            + CarbonTablePath.SEGMENT_EXT;
+    writeSegmentFile(segmentFile, writePath);
+    // Check whether we can completly remove the segment.
+    boolean deleteSegment = true;
+    for (Map.Entry<String, FolderDetails> entry : segmentFile.getLocationMap().entrySet()) {
+      if (entry.getValue().getStatus().equals(SegmentStatus.SUCCESS.getMessage())) {
+        deleteSegment = false;
+      }
+    }
+    if (deleteSegment) {
+      toBeDeletedSegments.add(segment.getSegmentId());
+    }
+    if (updateSegment) {
+      toBeUpdatedSegments.add(segment.getSegmentId());
+    }
+  }
+
+  /**
+   * Update the table status file with the dropped partitions information
+   *
+   * @param carbonTable
+   * @param uniqueId
+   * @param toBeUpdatedSegments
+   * @param toBeDeleteSegments
+   * @throws IOException
+   */
+  public static void commitDropPartitions(CarbonTable carbonTable, String uniqueId,
+      List<String> toBeUpdatedSegments, List<String> toBeDeleteSegments) throws IOException {
+    Set<Segment> segmentSet = new HashSet<>(
```

Comment:
Move segmentSet creation inside the below if check

Reply:
ok